Support GoTo for Go using `godef`

Manishearth opened the pull request:
fixes #333

Unfinished; I need to add the submodule (this currently works with a `godef_binary_path`), however for that rogpeppe/godef#24 must merge first (if not I can point it to my fork).

I also need to add a test.

Still, ready for incremental review; cc @Valloric

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/340)
<!-- Reviewable:end -->

Manishearth commented:
Added a test and the submodule. Submodule currently points to my fork; we can change that later.

Manishearth commented:
I forgot to hook it into the build command. Since gocode and godef both just need a `go build` to work (no additional deps for godef), I just made `--gocode-completer` build godef as well. Let me know if I should rename it or add another command.

vheon commented:
Reviewed 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 55 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YcY7R6HZS8f4GS3mo:-K9YcY7SppFZOO8x5f4Z:-268992707) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L55)):*
I think it would be sensible to change this variable while we're at it since now we have two binaries.

---

*[ycmd/completers/go/gocode_completer.py, line 207 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YdzyligGvbKvBlEos:-K9YdzyligGvbKvBlEot:-1250408191) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L207)):*
This function seems a little too convoluted; maybe would be better to split it into more helper functions.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

vheon commented:
Actually I think we have a plan to rename all the flag after the language instead of the backend, so I think that you can leave it like this.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9Yeqctb_V6FibnEuLV:-1470396501)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Thanks for doing this. This PR is looking pretty good :)

I've added some comments. Some of them pedantic, others worth a quick head scratch.

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 15 unresolved discussions.

---

*[README.md, line 237 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YeFWZHJBSt8AvoFs3:-K9YeFWZHJBSt8AvoFs4:860244663) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/README.md#L237)):*
This link seems to differ from the submodule path. Is that intentional? seems strange.

---

*[ycmd/completers/go/gocode_completer.py, line 134 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yejev1rkStYKMFo4h:-K9Yejev1rkStYKMFo4i:615536858) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L134)):*
2 blank lines between methods

---

*[ycmd/completers/go/gocode_completer.py, line 157 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YemCmcZmbYNQ4mSj2:-K9YemCnsLMgr0UpfFq2:1591447904) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L157)):*
2 blank lines

---

*[ycmd/completers/go/gocode_completer.py, line 191 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YenG9MPp9h3b2vOG5:-K9YenG9MPp9h3b2vOG6:1591447904) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L191)):*
2 blank lines

---

*[ycmd/completers/go/gocode_completer.py, line 192 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YerZzbOl-VskNzCrE:-K9YerZzbOl-VskNzCrF:-2084838224) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L192)):*
spaces inside `(` and `)`

---

*[ycmd/completers/go/gocode_completer.py, line 195 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfCUzlg58ehgSxo9h:-K9YfCUzlg58ehgSxo9i:527192792) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L195)):*
this seems to duplicate a log entry in `_GoToDefinition` - should it include different signpost text and the arguments?
Also info is probably too common a log level. This looks like debug

---

*[ycmd/completers/go/gocode_completer.py, line 196 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YgQwBiFd9yJh3SZSI:-K9YgQwBiFd9yJh3SZSJ:1111401455) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L196)):*
do we want to allow multiple requests simultaneously? each request runs in a separate thread, so we could be doing this in parallel. i don't think that's a problem, and unlikely from Vim as these commands are blocking, put worth thinking about.

---

*[ycmd/completers/go/gocode_completer.py, line 206 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yf7_GcWInrxx0BiCI:-K9Yf7_GcWInrxx0BiCJ:1591447904) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L206)):*
2 blank lines

---

*[ycmd/completers/go/gocode_completer.py, line 210 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfPCI7Bf_1EOKgFbd:-K9YfPCI7Bf_1EOKgFbe:-40846472) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L210)):*
i suspect this is debug output, rather than info

---

*[ycmd/completers/go/gocode_completer.py, line 217 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfZtoN-xN0-9zSZ3f:-K9YfZtoN-xN0-9zSZ3g:1046755980) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L217)):*
does this work if filename contains spaces or special/utf-8 chars ? Or is there some shell escaping magic required? Or does popen handle that?

---

*[ycmd/completers/go/gocode_completer.py, line 227 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfplwblUVbjZJgSgG:-K9YfplwblUVbjZJgSgH:-395294755) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L227)):*
line these up with `parsed[`

---

*[ycmd/tests/go/subcommands_test.py, line 3 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfwZqiOUTJHir2Rap:-K9YfwZqiOUTJHir2Raq:930638980) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L3)):*
it's 2016 now, right :)

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9Yg5Hx30uFIfBNrbJ4:-568408761) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
do these test cases really cover all of the code paths in _GoTo in the code? Like the one where godef returns just a filename, and not a column/line ? are there ways to test the failure cases?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9YgifPNaLfpDOsihGN:1569405092)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Done.

---

Review status: 5 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 195 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfCUzlg58ehgSxo9h:-K9Yi-L9MXot5zHPjFqH:1537660756) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L195)):*
Yeah, this was debug,  I removed it later

---

*[ycmd/completers/go/gocode_completer.py, line 196 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YgQwBiFd9yJh3SZSI:-K9Yi9ewpMSptePcPXD_:-1277637514) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L196)):*
I don't see any issue with it being in parallel. Godef is just a binary, if a user fires off two godef requests in rapid succession it should just open two new windows or whatever GoTo does for that editor.

---

*[ycmd/completers/go/gocode_completer.py, line 217 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfZtoN-xN0-9zSZ3f:-K9YiWh4WPerdRVC0SfY:110004995) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L217)):*
popen should handle it.

I tried it with a file with a space and it worked fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9YggtlWfQ9asi219cU:-499146009)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 5 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[README.md, line 237 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YeFWZHJBSt8AvoFs3:-K9Yk7kZ-7w8x3ES87TC:238428399) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/README.md#L237)):*
Yes. Manishearth/godef is a fork with a small fix that hasn't landed in rogpeppe/godef yet. I intend to update the submodule path when this happens.

---

*[ycmd/completers/go/gocode_completer.py, line 55 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YcY7R6HZS8f4GS3mo:-K9YkK-abgVBGZN4gdqy:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L55)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 207 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YdzyligGvbKvBlEos:-K9YkPc8N179IBTBi-Go:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L207)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 134 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yejev1rkStYKMFo4h:-K9YkKgGjuMsf8iwzrOZ:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L134)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 157 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YemCmcZmbYNQ4mSj2:-K9YkLuxHmNGcZvZfjtM:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L157)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 191 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YenG9MPp9h3b2vOG5:-K9YkMiKmtCrdfMxcu1A:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L191)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 192 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YerZzbOl-VskNzCrE:-K9YkNJJE961T5dY-VVg:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L192)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 206 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yf7_GcWInrxx0BiCI:-K9YkOzcJ8y-7QgHNPml:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L206)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 210 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfPCI7Bf_1EOKgFbd:-K9YjLnNigq6rRnSm002:1144958411) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L210)):*
We have an info() for ComputeCandidatesInner as well

---

*[ycmd/completers/go/gocode_completer.py, line 227 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfplwblUVbjZJgSgG:-K9YkRvHjL5tqJojQi_U:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/completers/go/gocode_completer.py#L227)):*
Done.

---

*[ycmd/tests/go/subcommands_test.py, line 3 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YfwZqiOUTJHir2Rap:-K9YkT26sdpmein7TaeA:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L3)):*
Done.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9YkUY3I4goGYTCLJ6H:1346922551) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
I'm not exactly familiar with the cases where godef will return a filename but not a line/col. It has to do with cases where a Go `Position` has a zero line. The Go parser (which godef uses) will be the one creating such `Position`s, but I don't know why it would. So far I haven't come across cases where this happens.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Alright, the build situation is a bit harder. Godef needs `$GOPATH` to be set up to work.

I can make some changes so that `go build` works locally in the submodule, however this will permanently need to use my fork in that case. Alternatively, I can modify the build process to `go install rogpeppe/godef` and use the binary from `$GOPATH` instead of a submodule.

What would you prefer?

Manishearth commented:
I went with the first solution. Now, `godef` is a submodule which can be build independently with `go build`.  However, the changes here are incompatible with regular-GOPATH installation, and can't be merged upstream, so we'll have to maintain a fork. (I'd still keep the URL in the readme pointing to the original godef, since the fork isn't much different semantically)


(Go packaging is pretty broken when it comes to choosing where you want to build/install things; you either must rigidly use `$GOPATH`, or use local paths outside of GOPATH, but the two are incompatible)

puremourning commented:
Review status: 4 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9YqVegrC89brTFSyUM:1662435232) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
Then is it dead code?

The reason I ask is that i'm not sure if passing 0,0 to BuildGoToResponse is even legit. ycmd uses 1-based indexes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 4 of 8 files reviewed at latest revision, 15 unresolved discussions, some commit checks broke.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9YqsCAo-qO0ew7LzkK:-928708789) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
I don't know; it's something that could happen.

Yeah, which is why I made it 1,0. But I think column is also 1-indexed, so I just made it 1,1. The idea is to jump to the first line in the file if we don't know the line/col.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Thanks for the PR! :) 

The implementation seems solid. Now that I've seen what integrating godef would look like, I'm fine with the idea.

The build situation though is very unfortunate. Could we add some env var hackery (temporarily set GOPATH to whatever) to our build.py script so that godef is built... _somewhere_ and then we just copy the final binary to where we want it? I don't know how much work that would be though. (Note that I don't really know anything about how Go is built since I don't use it so my suggestion may be nonsensical.)

We could live with depending on a fork, but I'd rather avoid it. 

---

Reviewed 1 of 2 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r7, 2 of 2 files at r8, 1 of 1 files at r9.
Review status: 7 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 55 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YcY7R6HZS8f4GS3mo:-K9Yr5xPHj6VErXxYH_a:-1362304515) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L55)):*
I'd rather name them `_binary_gocode` and `_binary_godef`. Eliding the `go` part is likely to confuse people (like it just confused me :)).

---

*[ycmd/completers/go/gocode_completer.py, line 204 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yrtq8o-KdrJVLtJOc:-K9Yrtq8o-KdrJVLtJOd:1744257796) ([raw file](https://github.com/valloric/ycmd/blob/996771cb33f111724c2b0b7d9382bf53d2ccebc5/ycmd/completers/go/gocode_completer.py#L204)):*
Instead of `proc`, maybe use `popen_handle`? Seems clearer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9YpiYvziUQHWFEd7B9:1928219101)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 1 of 1 files at r7, 1 of 2 files at r8, 1 of 1 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9Yrhu6oj-PyOnVmQ8D:-1661352752) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
I'm just nervous of untest(able|ed) code. Happy to take a view from other reviewers, but my view is that if you never executed a particular line of code, Schrodinger result applies - it might do what you expect... or it might launch a nuclear weapon :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
> The build situation though is very unfortunate. Could we add some env var hackery (temporarily set GOPATH to whatever) to our build.py script so that godef is built... somewhere

This is doable, but very hacky. Plus I'm not entirely certain we can rely on GOPATH to work a certain way cross platform. Additionally, we can no longer pin to a version like we can do with submodules. If the godef maintainer has a breaking change, all new installs will just be broken.

Valloric commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9YszyICg7jOQPXVbQr:1774508104) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
I agree with @puremourning here. I'm probably a bit more extreme on it than he is; I tend to view code without test coverage as a prime candidate for deletion (only _slightly_ exaggerating :)). If it's important, it should have tests. If it's hard to test, it should be rewritten until it's easier to test. :)

_Sigh_, ycmd itself doesn't have nearly as good test coverage as I'd like it to have. I need to work on improving that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 55 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YcY7R6HZS8f4GS3mo:-K9YtuY5b-CW8tVTn67b:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/5869c62af1df2b931cb1a4ef05cd0f610c69d3dc/ycmd/completers/go/gocode_completer.py#L55)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 204 \[r9\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yrtq8o-KdrJVLtJOc:-K9YtvWBh3u119AORDFi:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/996771cb33f111724c2b0b7d9382bf53d2ccebc5/ycmd/completers/go/gocode_completer.py#L204)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Valloric commented:
If we can't pin to a version, that's a dealbreaker, naturally. I'd rather go with a fork.

---

Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9YtwmTUr8J_vrgCxSe:1118289003)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/tests/go/subcommands_test.py, line 49 \[r5\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yg5Hx30uFIfBNrbJ3:-K9YuQO1As3A9MbZEJDU:-1153910665) ([raw file](https://github.com/valloric/ycmd/blob/b23e3e8e51a3a53dc4202e5930532a41331c57b0/ycmd/tests/go/subcommands_test.py#L49)):*
I deleted the code which tries to handle filename-only responses. Incomplete responses from godef will now be discarded.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 2 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r7, 1 of 2 files at r8, 1 of 1 files at r9, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 60 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yy7CIl5KpIyA8sEuX:-K9Yy7CJaKSSX2GbR1um:1703793557) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L60)):*
You should use a placeholder in `BINARY_NOT_FOUND_MESSAGE`:
```python
BINARY_NOT_FOUND_MESSAGE = '{0} binary not found'

_logger.error( BINARY_NOT_FOUND_MESSAGE.format( 'godef' ) )
```

---

*[ycmd/completers/go/gocode_completer.py, line 163 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yv_XAJVyxDs9hSNjR:-K9Yv_XAJVyxDs9hSNjS:-649042624) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L163)):*
`FindGocodeBinary` and `FindGoDefBinary` are near identicals. They should be merged into one method `FindBinary` with additional parameters.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YwUN2j8c0g553sEwg:-K9YwUN2j8c0g553sEwh:441930458) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L213)):*
Same comment as above for these two methods. I would merge them into the following method:
```python
_ExecuteBinary( self, args, contents = None )
```
with `args` being a list (`_popener` already takes a list so we can directly pass `args` this way).

`COMPLETION_ERROR_MESSAGE` needs to be renamed.

---

*[ycmd/completers/go/gocode_completer.py, line 239 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yx3xKUgepgSBpA06D:-K9Yx3xLGnEtIynJHgW0:730621216) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L239)):*
`else` branch is not needed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 7 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 60 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yy7CIl5KpIyA8sEuX:-K9Z3suHimcijI_NELAs:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L60)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 163 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yv_XAJVyxDs9hSNjR:-K9Z3ro89h4nP6FeJETG:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L163)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9YwUN2j8c0g553sEwg:-K9Z3n9lhMmlKM_t5qKw:-1569853652) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L213)):*
Done. I kept it as *args, **kwargs though.

---

*[ycmd/completers/go/gocode_completer.py, line 239 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yx3xKUgepgSBpA06D:-K9Z3h041JGBFI3BTqRT:-2026750072) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L239)):*
How else should I bubble up to the try block above it?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 239 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yx3xKUgepgSBpA06D:-K9Z4h5_r_LHBt9q6hVr:-401425811) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L239)):*
Sorry, I mean that:
```python
if ...
  return ...
raise ...
```
`else` is not necessary.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 239 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yx3xKUgepgSBpA06D:-K9Z59-iVLcrP7EjtT3E:1987845459) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L239)):*
or maybe better, you could do:
```python
if not ...
  raise ...
return ...
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 239 \[r12\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Yx3xKUgepgSBpA06D:-K9Z5VYv8qJpfBq0H4wt:1890975904) ([raw file](https://github.com/valloric/ycmd/blob/bd41d93da2e4e5c09a144b393bd9a11f25a9c9c7/ycmd/completers/go/gocode_completer.py#L239)):*
Done. `if not` makes the condition even more complicated, I prefer the first one

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[build.py, line 288 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z6dfmx-sR_xY7NVTr:-K9Z6dfmx-sR_xY7NVTs:-589274487) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/build.py#L288)):*
Two blank lines around function.

---

*[ycmd/completers/go/gocode_completer.py, line 62 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z70EEkRirMAl6A36J:-K9Z70EEkRirMAl6A36K:-500982111) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L62)):*
Line over 80 characters.

---

*[ycmd/completers/go/gocode_completer.py, line 68 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7AXgbGsheXxnlGaW:-K9Z7AXgbGsheXxnlGaX:-500982111) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L68)):*
Line over 80 characters.

---

*[ycmd/completers/go/gocode_completer.py, line 179 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7Lfn2d9FkwZUwfbA:-K9Z7Lfn2d9FkwZUwfbB:-500982111) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L179)):*
Line over 80 characters.

---

*[ycmd/completers/go/gocode_completer.py, line 180 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7FYv39yknk0eRa1C:-K9Z7FYv39yknk0eRa1D:1160233803) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L180)):*
`popen_handle.returncode` instead of `proc.returncode`.

---

*[ycmd/completers/go/gocode_completer.py, line 196 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7bnyiovxYOYc2op0:-K9Z7bnyiovxYOYc2op1:-498485081) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L196)):*
Trailing space.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 6 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[build.py, line 288 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z6dfmx-sR_xY7NVTr:-K9Z9mbfsVzrkEz72O9Y:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/build.py#L288)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 62 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z70EEkRirMAl6A36J:-K9Z9iEDg2J1OSXp7zIm:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L62)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 68 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7AXgbGsheXxnlGaW:-K9Z9iYDTC8hVuXcGXNn:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L68)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 179 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7Lfn2d9FkwZUwfbA:-K9Z9jl7Ot63lyrEJzr5:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L179)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 180 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7FYv39yknk0eRa1C:-K9Z9iwkOKzBbG_yb-Kd:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L180)):*
Done.

---

*[ycmd/completers/go/gocode_completer.py, line 196 \[r14\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9Z7bnyiovxYOYc2op0:-K9Z9jZQe_FsI4IbXgiP:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/e5e90bd5cf6a2f8c25b5c95e3f8bcadcd5345971/ycmd/completers/go/gocode_completer.py#L196)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I'm <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> on this (others might not be yet). Great work! :)

---

Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ZCVJinc3vELp-oRJ7:-12422120)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Oh, and please do squash down the commits once you're ready to have this merged.

---

Reviewed 1 of 2 files at r15, 1 of 1 files at r16.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ZG1rVlOwh8MUdHa-U:-990951744)*
<!-- Sent from Reviewable.io -->

micbou commented:
I am getting this error when building `godef`:
```
acme.go:11:2: cannot find package "9fans.net/go/acme" in any of:
        C:\Go\src\9fans.net\go\acme (from $GOROOT)
        C:\Users\micbou\go\src\9fans.net\go\acme (from $GOPATH)
```
It happens also on Travis and AppVeyor.

---

Reviewed 2 of 2 files at r15.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ZGgJfT4WHHSUTbvnQ:-157017153)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Oh, gopath strikes again. I'll fix this tomorrow.

Manishearth commented:
@micbou can you try with the latest update?

micbou commented:
It builds fine now. Maybe this is expected but I find the `GoTo` behavior a little strange. For example:
```go
package main

type Struct struct {
    member int
}
```
When on `member`, `GoTo` does not always go to the start of the word (it returns an error on the first 3 characters, jump for the other characters). 

---

Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ZMcPbAQw-D-sLj5QM:1341047824)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Build passes now.

Manishearth commented:
@vheon @puremourning All proposed changes made. Could I get review on the final commits?

Manishearth commented:
@micbou This works for me.

micbou commented:
I am getting this issue because of Windows newlines (`\r\n`). Offset returned by `_ComputeOffset` does not correspond to the file contents in this case. But that's not all, try to jump in a unsaved buffer. It will jump to the wrong location. This is because `godef` use the contents from the file instead of the current contents. We need to directly pass the contents to `godef` using the `-i` option. Adding a test for this would be a good idea (using a file with Windows newlines for example).

---

Reviewed 1 of 1 files at r16, 1 of 1 files at r17, 2 of 2 files at r18.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r18\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9aDnVizWJ3LlA1BYUy:-K9aDnVjC17OeuLIeVF5:-1397982027) ([raw file](https://github.com/valloric/ycmd/blob/9edcb5798388a47552aac40882944ed801134097/ycmd/completers/go/gocode_completer.py#L213)):*
Spaces after `(` and before `)`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9aDymVB-6qfbuf2SJ_:723637291)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
I modified gocode to support unsaved buffers and updated ycmd to work with this.

I created a file with windows line endings and it worked; perhaps it was the buffering issue that was affecting you? ComputeOffset is used by all the completers so should work on windows.

---

Review status: 7 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r18\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9aDnVizWJ3LlA1BYUy:-K9b2JhHAFNqsoJAEnVc:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/9edcb5798388a47552aac40882944ed801134097/ycmd/completers/go/gocode_completer.py#L213)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9b2LqnsF_dQzemwxqu:-1264741280)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Seems to pass Travis now, too.

micbou commented:
> I created a file with windows line endings and it worked; perhaps it was the buffering issue that was affecting you? 

Yes, same cause. This is why I suggested to use Windows line endings as a test.

> ComputeOffset is used by all the completers

If by all, you mean only the Go completer then I agree.

---

Reviewed 2 of 2 files at r19, 2 of 2 files at r20, 1 of 1 files at r21.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r19\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9b56qWu4V6toHIaIt4:-K9b56qWu4V6toHIaIt5:-429785601) ([raw file](https://github.com/valloric/ycmd/blob/d4286540dc236ff4e0f607a657f58a89d6ce577b/ycmd/completers/go/gocode_completer.py#L213)):*
Why did you make this change in `godef`? It was working fine by passing both `-f` and `-i` flags. Also, `-` is a valid filename so it would be better to return `null` or something like that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9b4cyG1KjDEJMZNgHx:-1249697076)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
> If by all, you mean only the Go completer then I agree.

Huh, I thought I'd taken that code from the Rust one.

The only case where ComputeOffset won't work is if it's a windows file that contains stray `\n`s.  Is this a case we care to handle?  (I'd prefer to do it in a followup if so; since it's not a bug in this PR per se, rather one in the existing gocode integration)

---

Review status: 8 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/go/gocode_completer.py, line 213 \[r19\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9b56qWu4V6toHIaIt4:-K9d4qUBPK1KXyr1973r:-1713942910) ([raw file](https://github.com/valloric/ycmd/blob/d4286540dc236ff4e0f607a657f58a89d6ce577b/ycmd/completers/go/gocode_completer.py#L213)):*
Oh, I didn't know that that worked. `-` isn't a valid filename, though; filenames are full paths. Will fix.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9d526371HCFkpr2A16:857284566)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
It is possible (takes some work, not much) to obviate the need for ComputeOffset in godef, btw, since godef takes in an offset and quickly converts it to an AST node. However, gocode still needs ComputeOffset since the use of a cursor offset is well embedded in its core.

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> after squashing the commits.

> The only case where ComputeOffset won't work is if it's a windows file that contains stray \ns. Is this a case we care to handle?

I think this is fixed by using the buffer contents instead of the file contents.

---

Reviewed 2 of 2 files at r22.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9i4K2Tb9KHQMbQyuar:1036602630)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Squashed

vheon commented:
Made a couple of comments. Other than that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r7, 1 of 2 files at r18, 1 of 2 files at r20, 1 of 2 files at r22.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 237 \[r22\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9iYiwtWFIaNHr3pkHX:-K9iYiwtWFIaNHr3pkHY:835671224) ([raw file](https://github.com/valloric/ycmd/blob/f538165289b972e9b3aa3b8acc72f433f28820f6/README.md#L237)):*
Just to be clear: are we going to ship your fork of `godef`? Is it a temporary thing due to https://github.com/rogpeppe/godef/pull/24 or is it also for the acme support which we don't want? I'm OK shipping your fork (might be easier to overcome bugs if we find them) but then you should change this link.

---

*[ycmd/completers/go/gocode_completer.py, line 201 \[r22\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9iY5G5XyrhhSVE2gKm:-K9iY5G6NTYOeIyuEO5O:540681775) ([raw file](https://github.com/valloric/ycmd/blob/f538165289b972e9b3aa3b8acc72f433f28820f6/ycmd/completers/go/gocode_completer.py#L201)):*
isn't the same writing

```python
"-f=", filename,
```
?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9iYona69hSNS5yLiAb:43808086)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Amended the last commit to fix the README.

Is there anything left for me to do to get this merged? :smile:

---

Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 237 \[r22\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9iYiwtWFIaNHr3pkHX:-K9ibKuGU1ehAFRQpm46:1067560119) ([raw file](https://github.com/valloric/ycmd/blob/f538165289b972e9b3aa3b8acc72f433f28820f6/README.md#L237)):*
It's a permanent thing for the following reasons:

 - We cannot pin to a version otherwise
 - We cannot build in-tree without permanent modifications which are incompatible with (and thus won't be merged in) regular godef due to the GOPATH dichotomy

I was linking to the source repo since that's where the actual engine lives; the fork only has compat changes. But I'll like to mine since you feel otherwise :smile:

---

*[ycmd/completers/go/gocode_completer.py, line 201 \[r22\]](https://reviewable.io:443/reviews/valloric/ycmd/340#-K9iY5G5XyrhhSVE2gKm:-K9ib1sL5aA5k1nhHBrL:682251842) ([raw file](https://github.com/valloric/ycmd/blob/f538165289b972e9b3aa3b8acc72f433f28820f6/ycmd/completers/go/gocode_completer.py#L201)):*
No, that will introduce a space between the arguments, which doesn't work.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ic-A6tPH_IVi-WA3c:459971602)*
<!-- Sent from Reviewable.io -->

micbou commented:
Excellent work.

@homu r+

---

Reviewed 1 of 1 files at r23.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/340#-:-K9ivQeZ8q9h-GE1AxCQ:931298648)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 24b43f9 has been approved by `micbou`

<!-- @homu r=micbou 24b43f9e951ff425165cc289585780c856cbe4b0 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/107118850)

Valloric commented:
@Manishearth Thanks for working on this! :)
